sim.1
.\"	This file is part of the software similarity tester SIM.
.\"	Written by Dick Grune, Vrije Universiteit, Amsterdam.
.\"	$Id: sim.1,v 2.28 2015-01-28 16:46:42 dick Exp $
.\"
.TH SIM 1 2015/01/23
.SH NAME
sim \- find similarities in C, Java, Pascal, Modula-2, Lisp, Miranda, or text files
.SH SYNOPSIS
.B sim_c
[
.B \-[defFiMnpPRsSTv]
.B \-r
.I N
.B \-t
.I N
.B \-w
.I N
.B \-o
.I F
]
file ... [ [
.B /
.B |
] file ... ]
.br
.B sim_c
\&...
.br
.B sim_java
\&...
.br
.B sim_pasc
\&...
.br
.B sim_m2
\&...
.br
.B sim_lisp
\&...
.br
.B sim_mira
\&...
.br
.B sim_text
\&...
.br
.SH DESCRIPTION
.I Sim_c
reads the C files
.I file ...
and looks for segments of text that are similar; two segments of program text
are similar if they only differ in layout, comment, identifiers, and
the contents of numbers, strings and characters.
If any runs of sufficient length
are found, they are reported on standard output; the number of significant
tokens in the run is given between square brackets.
.PP
.I Sim_java
does the same for Java,
.I sim_pasc
for Pascal,
.I sim_m2
for Modula-2,
.I sim_mira
for Miranda,
and
.I sim_lisp
for Lisp.
.I Sim_text
works on arbitrary text and it is occasionally useful on shell scripts.
.PP
The program can be used for finding copied pieces of code in
purportedly unrelated programs (with
.B \-s
or
.BR \-S ),
or for finding accidentally duplicated code in larger projects (with
.B \-f
or
.BR \-F ).
.PP
If a separator
.B /
or
.B |
is present in the list of input files, the files are divided into a group of
"new" files (before the
.BR /
or
.BR | )
and a group of "old" files; if there is no
.BR /
or
.BR | ,
all files are "new".
Old files are never compared to each other.
See also the description of the
.B \-s
and
.B \-S
options below.
.PP
Since the similarity tester needs file names to pinpoint the similarities, it
cannot read from standard input.
.PP
There are the following options:
.TP
.B \-d
The output is in a diff(1)-like format instead of the default
2-column format.
.TP
.B \-e
Each file is compared to each file in isolation; this will find all
similarities between all texts involved, regardless of repetitive text
(see `Calculating Percentages' below).
.TP
.B \-f
Runs are restricted to segments with balancing parentheses, to isolate
potential routine bodies (not in
.IR sim_text ).
.TP
.B \-F
The names of routines in calls are required to match exactly
(not in
.IR sim_text ).
.TP
.B \-i
The names of the files to be compared are read from standard input, including
a possible separator
.BR /
or
.BR | ;
the file names must be one to a line.
This option allows a very large number of file names to be specified;
it differs from the \fC@\fP facility provided by some compilers in that it
handles file names only, and does not recognize option arguments.
.TP
.B \-M
Memory usage information is displayed on standard error output.
.TP
.B \-n
Similarities found are summarized by file name, position and size, rather than
displayed in full.
.TP
.B "\-o F"
The output is written to the file named
.IR F .
.TP
.B \-p
The output is given in similarity percentages; see `Calculating Percentages'
below; implies \fB\-e\fP and \fB\-s\fP.
.TP
.B \-P
As
.B \-p
but only the main contributor is shown; implies \fB\-e\fP and \fB\-s\fP.
.TP
.B "\-r N"
The minimum run length is set to
.I N
units; the default is 24 tokens, except in
.IR sim_text ,
where it is 8 words.
.TP
.B \-R
Directories in the input list are entered recursively, and all files they
contain are involved in the comparison.
.TP
.B \-s
The contents of a file are not compared to itself (\-s for "not self").
.TP
.B \-S
The contents of the new files are compared to the old files only \- not
between themselves.
.TP
.B "\-t N"
In combination with the
.B \-p
or
.B \-P
options, sets the threshold (in percent) below which similarities will not be
reported; the default is 1, except in
.IR sim_text ,
where it is 20.
.TP
.B \-T
A more terse and uniform form of output is produced, which may be more
suitable for postprocessing.
.TP
.B \-v
Prints the version number and compilation date on standard output, then stops.
.TP
.B "\-w N"
The page width used is set to
.I N
columns; the default is 80.
.TP
.B "\-\-"
(A secret option, which prints the input as the similarity checker sees it,
and then stops.)
.PP
The
.B \-p
option results in lines of the form
.nf
.ft C
        F consists for x % of G material
.ft P
.fi
meaning that \fCx\fP % of \fCF\fP's text can also be found in \fCG\fP.
Note that this relation is not symmetric; it is in fact quite possible for one
file to consist for 100 % of text from another file, while the other file
consists for only 1 % of text of the first file, if their lengths differ
enough.
The
.B \-P
(capital P) option shows the main contributor for each file only.
This simplifies the identification of a set of files \fCA[1] ... A[n]\fP,
where the concatenation of these files is also present.
A threshold can be set using the
.B \-t
option; note that the granularity of the recognized text is still governed by
the
.B \-r
option or its default.
.PP
The
.B \-r
option controls the number of "units" that constitute a run.
For the programs that compare programming language code, a unit is a lexical
token in the pertinent language; comment and standard preamble material (file
inclusion, etc.) is ignored and all strings are considered the same.
For
.I sim_text
a unit is a "word" which is defined as any sequence of one or more letters,
digits, or characters over 127 (177 octal), (to accommodate letters such as
\(:a, \(/o, etc.).
.br
.I Sim_text
accepts  s p a c e d   t e x t  as normal text.
.PP
The
.B \-s
and
.B \-S
options control which files to compare.
Input files are divided into two groups, new and old.
In the absence of these control options the programs compare the files thus
(for 4 new files and 6 old ones):
.nf
.ft C
                          n e w    /     o l d       <- first file
                        1  2  3  4 / 5  6  7  8  9 10
                      |------------/------------
                 n  1 | c          /
                 e  2 | c  c       /
                 w  3 | c  c  c    /
                    4 | c  c  c  c /
       second       / / /  /  /  / / /  /  /  /  /  /
       file  ->     5 | c  c  c  c /
                 o  6 | c  c  c  c /
                 l  7 | c  c  c  c /
                 d  8 | c  c  c  c /
                    9 | c  c  c  c /
                   10 | c  c  c  c /
.ft P
.fi
where the \fCc\fPs represent file comparisons, and the \fC/\fP the
demarcation between new and old files.
.bp					\" KLUGDE ZZ
Using the
.B \-s
option reduces this to:
.nf
.ft C
                          n e w    /     o l d       <- first file
                        1  2  3  4 / 5  6  7  8  9 10
                      |------------/------------
                 n  1 |            /
                 e  2 | c          /
                 w  3 | c  c       /
                    4 | c  c  c    /
       second       / / /  /  /  / / /  /  /  /  /  /
       file  ->     5 | c  c  c  c /
                 o  6 | c  c  c  c /
                 l  7 | c  c  c  c /
                 d  8 | c  c  c  c /
                    9 | c  c  c  c /
                   10 | c  c  c  c /
.ft P
.fi
The
.B \-S
option reduces this further to:
.nf
.ft C
                          n e w    /     o l d       <- first file
                        1  2  3  4 / 5  6  7  8  9 10
                      |------------/------------
                 n  1 |            /
                 e  2 |            /
                 w  3 |            /
                    4 |            /
       second       / / /  /  /  / / /  /  /  /  /  /
       file  ->     5 | c  c  c  c /
                 o  6 | c  c  c  c /
                 l  7 | c  c  c  c /
                 d  8 | c  c  c  c /
                    9 | c  c  c  c /
                   10 | c  c  c  c /
.ft P
.fi
.PP
The programs can handle UNICODE file names under Windows.
This is relevant only under the
.B \-R
option, since there is no way to give UNICODE file names from the command line.
.SH LIMITATIONS
Repetitive input is the bane of similarity checking.
If we have a file containing 4 copies of identical text,
.nf
.ft C
    A1 A2 A3 A4
.ft P
.fi
where the numbers serve only to distinguish the identical copies,
there are 8 identities: \fCA1=A2\fP, \fCA1=A3\fP, \fCA1=A4\fP, \fCA2=A3\fP, \fCA2=A4\fP, \fCA3=A4\fP, \fCA1A2=A3A4\fP,
and \fCA1A2A3=A2A3A4\fP.
Of these, only 3 are meaningful: \fCA1=A2\fP, \fCA2=A3\fP, and \fCA3=A4\fP.
And for a table with 20 lines identical to each other, not unusual in a program,
there are 715 identities, of which at most 19 are meaningful.
Reporting all 715 of them is clearly unacceptable.
.PP
To remedy this, finding the identities is performed as follows:
For each position in the text, the largest segment is found, of which a
non-overlapping copy occurs in the text following it.
That segment and its copy are reported and scanning resumes at the position
just after the segment.
For the above example this results in the identities \fCA1A2=A3A4\fP and \fCA3=A4\fP,
which is quite satisfactory, and for \fIN\fP identical segments roughly
\fI2 log N\fP messages are given.
.PP
This also works out well when the four identical segments are in different
files:
.nf
.ft C
    File1: A1
    File2: A2
    File3: A3
    File4: A4
.ft P
.fi
Now combined segments like \fCA1A2\fP do not occur, and the algorithm finds
the runs \fCA1=A2\fP, \fCA2=A3\fP, and \fCA3=A4\fP, for a total of \fIN-1\fP
runs, all informative.
.SS Calculating Percentages
The above approach is not suitable for obtaining the percentage of a file's
content that can be found in another file.
This requires comparing in isolation each file pair represented by a \fCc\fP
in the matrixes above; this is what the \fB\-e\fP option does.
Under the \fB\-e\fP option a segment \fCFile1:A1\fP, recognized in \fCFile2\fP, will again
be recognized in \fCFile3\fP and \fCFile4\fP.
In the example above it produces the runs
.nf
.ft C
    File1:A1=File2:A2
    File1:A1=File3:A3
    File1:A1=File4:A4
    File2:A2=File3:A3
    File2:A2=File4:A4
    File3:A3=File4:A4
.ft P
.fi
for a total of \fI\(12N(N-1)\fP runs.
.SH TIME AND SPACE REQUIREMENTS
Care has been taken to keep the time requirements of all internal processes
(almost) linear in the lengths of the input files, by using various tables.
If, however, there is not enough memory for the tables, they are discarded in
order of unimportance, under which conditions the algorithms revert to their
quadratic nature.
.PP
The time requirements are quadratic in the number of files.
This means that, for example, one 64 MB file processes much faster than 8000 8
kB files.
.PP
The program requires 6 bytes of memory for each token in the input; 2
bytes per newline (not when doing percentages); and about 76 bytes for each
run found.
.SH EXAMPLES
The call
.nf
.ft C
        sim_c *.c
.ft P
.fi
highlights duplicate code in the directory.
(It is useful to remove generated files first.)
A call
.nf
.ft C
        sim_c -f -F *.c
.ft P
.fi
can pinpoint them further.
.PP
A call
.nf
.ft C
        sim_text -e -p -s new/* / old/*
.ft P
.fi
compares each file in \fCnew/*\fP to each file in \fCnew/*\fP and
\fCold/*\fP, and if any pair has more that 20% in common, that fact is
reported.
Usually a similarity of 30% or more is significant; lower than 20% is probably
coincidence; and in between is doubtful.
.PP
A call
.nf
.ft C
        sim_text -e -n -s -r100 new/* "|" old/*
.ft P
.fi
compares the same files, and reports large common segments.
(The
.B |
can be used as a separator instead of
.B /
on systems where the
.B /
as a command-line parameter gets mangled by the command interpreter.)
.PP
Both approaches are good for plagiarism detection.
.SH BUGS
Since it uses
.I lex(1)
on some systems, it may crash on any weird construction in the input that
overflows
.IR lex 's
internal buffers, for example an identifier of several thousand letters long.
.SH AUTHOR
Dick Grune, Vrije Universiteit, Amsterdam; dick@dickgrune.com.


ChangeLog
2014-02-17  Dick Grune  <Gebruiker@GEBRUIKER-PC>

	* sim.c (is_new_old_separator): MinGW sometimes (?) interprets the /
	as a command-line argument as a reference to the MinGW tree, which
	makes the / unusable as a separator. Even escaping it ("/") does not
	help. Added the | as a separator.

2014-01-26  Dick Grune

	*  %z from Marcus Brinkmann implemented by a routine size_t2string.

2014-01-26  Dick Grune

	*  Sizes on my present machine (HP 6730b laptop):
	unsigned short int: 2
	unsigned int: 4
	unsigned long int: 4
	unsigned long long int: 8

2013-05-31  Dick Grune

	*  (hash.c) Better hash2() function.

2013-04-28  Dick Grune

	*  Markus Brinkmann (marcus.brinkmann@ruhr-uni-bochum.de) supplied a
	64-bit version. %z does not work on MinGW C.

2012-11-28  Dick Grune

	* newargs.c (recursive_args): Liqun Chen (liqun.chen@hp.com)
	submitted a bug report noting that the separator / is expanded under
	the -R option. Corrected.

2012-09-30  Dick Grune

	* pass2.c (pass2_txt): Boyd Blackwell (Boyd.Blackwell@anu.edu.au)
	submitted a bug report in which the line numbers (and runs
	representations) were way off (75 lines). The input files were
	characterized by extremely long lines, hundreds of tokens (max. 521).

	After 2.5 days of debugging the cause was found: 1. since the mapping
	from token positions to line numbers is stored as the difference of
	the token positions from one line to the next (see text.c); 2. since
	these differences are stored in unsigned chars to save space; 3. since
	the nl_buff mechanism is switched off when one of these unsigned
	characters overflow; and since 521 tokens on one line overflowed this
	unsigned char, the nl_buff mechanism was shut off.

	Since when there is no nl_buff information in pass2, pass2 resorts to
	rereading the input file calling yylex again; 2. since the preceding
	file had few runs to find line number to, the preceding file was not
	read to the end, and the rest remained in flex's buffer, so a portion
	of the preceding file seemed prefixed to the present file, adding 75
	lines to it.

	Remedy: flushing flex's buffer explicitly in pass2_txt(); this is
	simpler than using flex's YY_BUFFER_STATE mechanism.

	Advice: get rid of the nl_buff mechanism; it is no longer relevant.

2012-06-09  Dick Grune

	* lang.h:
	The *lang.l files are unusual in two respects:
	1. they present two interfaces to the rest of the system:
	language.[ch], static data about the language, and lang.[ch], dynamic
	data about the input file's content;
	2. both interfaces come with multiple implementations, one for each
	*lang.l file; i.e., they are abstract.
	This has been sorted out with some difficulty.

2012-05-08  Dick Grune

	* Changed to 16-bit tokens, for better resolution for sim_text and
	on -F option, and for UTF-8 input.

	It was not worth while to save the 8-bit token code: on serious
	comparisons the increase in memory usage is about 10% (330 000 on a
	maximum allocation of 3 030 976 for comparing the sources of MCD2).

2009-03-11  Dick Grune  <dick@flits.few.vu.nl>

	* newargs.c: added -R option to follow directories recursively.
	  See recursive_args().

2008-09-22    <Dick@ACER>
	* added newargs.[ch], to supply file names from standard input,
	  for those compilers that do not have the @ facility. Implemented
	  without fixed limits.

2008-09-21    <Dick@ACER>
	* changed default format back to original, and inverted the
	  -v(erbose) option into a -T(erse) option.

2008-03-31  Dick Grune  <dick@flits.few.vu.nl>
	* *.l: the following are not universally recognized; removed.
		%option nounput
		%option never-interactive

2008-03-31    <Dick@ACER>
	Introduced aiso.* and Malloc.? as imported modules.

2007-11-21  Carlos Maziero <maziero@ppgia.pucpr.br>
	- output format modified in order to facilitate "grep" filtering
	- added option "-v" for a more verbose output
	- added option "-tN" to define a threshold %N (only similarities
	  over N% are shown)
	- fixed SEGV on writing to the output file
	- the file list can be informed through STDIN (one file per line,
	  accepts "/" marker); this is useful for compilers that lack the
	  @ facility

2007-08-23  Dick Grune  <dick@hydra.cs.vu.nl>
	LICENSE.txt added.

2006-11-27  Dick Grune  <dick@hydra.cs.vu.nl>
	Removal of setbuff() for compatibility.

2005-01-17  Dick Grune  <dick@blade014.cs.vu.nl>
	Corrections by Jerry James <james@eecs.ku.edu>; ANSIizing, etc.

2004-08-05  Dick Grune  <dick@blade014.cs.vu.nl>
	Finished the 'percentage' option.

08-Nov-2001	Dick Grune
	Begun to add a 'percentage' option, which will express the
	similarity between two files in percents.

27-Sep-2001	Dick Grune
	Split add_run() off from compare.c into add_run.c, to accommodate
	different add_run()s, for different types of processing.

27-Nov-1998	Dick Grune
	Installed a Miranda version supplied by Emma Norling (ejn@cs.mu.oz.au)

23-Feb-1998	Dick Grune
	Renamed text.l to textlang.l for uniformity and to make room for
	a possible module text.[ch].

	Isolated a module for handling the token array from buff.[ch] to
	tokenarray.[ch], and renamed buff.[ch] to text.[ch].

23-Feb-1998	Dick Grune
	There is probably not much point in abandoning the nl_buff list
	when running out of memory for TokenArray[]: each token costs 1
	byte for the token and 4 bytes for the entry in
	forward_references[], a total of 5 bytes.  There are about 3
	tokens to a line, together requiring 15 bytes, plus 1 byte in
	nl_buff yields 16 bytes.  So releasing nl_buff frees only 1/16 =
	6.7 % of memory.

	Since the code is a bother, I removed it.  Note that nl_buff is
	still abandoned when the number of tokens in a line does not fit
	in one unsigned char (but that is not very likely to happen).


21-Feb-1998	Dick Grune
	Printing got into an infinite loop when the last line of the
	input was not terminated by a newline AND contained tokens that
	were included in a matching run.
	This was due to a double bug: 1. the non-terminated line was not

sim.c
/*	This file is part of the software similarity tester SIM.
	Written by Dick Grune, Vrije Universiteit, Amsterdam.
	$Id: sim.c,v 2.45 2015-04-29 18:18:22 dick Exp $
*/

#include	<stdio.h>
#include	<stdlib.h>
#include	<string.h>

#include	"system.par"
#include	"settings.par"
#include	"sim.h"
#include	"options.h"
#include	"newargs.h"
#include	"token.h"
#include	"language.h"
#include	"error.h"
#include	"text.h"
#include	"runs.h"
#include	"hash.h"
#include	"compare.h"
#include	"pass1.h"
#include	"pass2.h"
#include	"pass3.h"
#include	"percentages.h"
#include	"stream.h"
#include	"lang.h"

#include	"Malloc.h"
#include	"any_int.h"

							/* VERSION */
#if	0	/* set to 1 when experimenting */
#undef	VERSION
#define	VERSION	__TIMESTAMP__
#endif
							/* PARAMETERS */
/* command-line parameters */
int Min_Run_Size = DEFAULT_MIN_RUN_SIZE;
int Page_Width = DEFAULT_PAGE_WIDTH;
int Threshold_Percentage = 1;		/* minimum percentage to show */
FILE *Output_File;
FILE *Debug_File;

/* and their string values, for language files that define their own parameters
*/
const char *token_name = "token";
const char *min_run_string;
const char *threshold_string;

const char *progname;			/* for error reporting */

static const char *page_width_string;
static const char *output_name;		/* for reporting */

static const struct option optlist[] = {
	{'r', "minimum run size", 'N', &min_run_string},
	{'w', "page width", 'N', &page_width_string},
	{'f', "function-like forms only", ' ', 0},
	{'F', "keep function identifiers in tact", ' ', 0},
	{'d', "use diff format for output", ' ', 0},
	{'T', "terse output", ' ', 0},
	{'n', "display headings only", ' ', 0},
	{'p', "use percentage format for output", ' ', 0},
	{'P', "use percentage format, main contributor only", ' ', 0},
	{'t', "threshold level of percentage to show", 'N', &threshold_string},
	{'e', "compare each file to each file separately", ' ', 0},
	{'s', "do not compare a file to itself", ' ', 0},
	{'S', "compare new files to old files only", ' ', 0},
	{'R', "recurse into subdirectories", ' ', 0},
	{'i', "read arguments (file names) from standard input", ' ', 0},
	{'o', "write output to file F", 'F', &output_name},
	{'v', "show version number and compilation date", ' ', 0},
	{'M', "show memory usage info", ' ', 0},
	{'-', "lexical scan output only", ' ', 0},
	{0, 0, 0, 0}
};

static void
allow_at_most_one_out_of(const char *opts) {
	const char *first;
	for (first = opts; *first; first++) {
		const char *second;
		for (second = first + 1; *second; second++) {
			if (is_set_option(*first) &&is_set_option(*second)) {
				char msg[256];
				sprintf(msg,
					"options -%c and -%c are incompatible",
					*first, *second
				);
				fatal(msg);
			}
		}
	}
}

							/* SERVICE ROUTINES */
int
is_new_old_separator(const char *s) {
	if (strcmp(s, "/") == 0) return 1;
	if (strcmp(s, "|") == 0) return 1;
	return 0;
}

const char *
size_t2string(size_t s) {
	return any_uint2string(s, 0);
}

							/* PROGRAM */
static void
read_and_compare_files(int argc, const char **argv, int round) {
	Read_Input_Files(argc, argv, round);
	Make_Forward_References();
	Compare_Files();
	Free_Forward_References();
}

#ifdef	ARG_TEST
static void
show_args(const char *msg, int argc, const char *argv[]) {
	fprintf(stdout, "%s: ", msg);

	int i;
	for (i = 0; i < argc; i++) {
		fprintf(stdout, "arg[%d] = %s; ", i, argv[i]);
	}
	fprintf(stdout, "\n");
}
#endif	/* ARG_TEST */

int
main(int argc, const char *argv[]) {

	/* Save program name */
	progname = argv[0];
	argv++, argc--;				/* and skip it */

	/* Set the default output and debug streams */
	Output_File = stdout;
	Debug_File = stdout;

	/* Get command line options */
	{	int nop = do_options(progname, optlist, argc, argv);
		argc -= nop, argv += nop;	/* and skip them */
	}

	/* Check options compatibility */
	allow_at_most_one_out_of("dnpPT");
	if (is_set_option('t')) {
		/* threshold means percentages */
		if (!is_set_option('p') && !is_set_option('P'))
			fatal("option -t requires -p or -P");
	}

	/* Treat the simple options */
	if (is_set_option('v')) {
		fprintf(stdout, "Version %s\n", VERSION);
		return 0;
	}

	if (is_set_option('P')) {
		set_option('p');
	}
	if (is_set_option('p')) {
		set_option('e');
		set_option('s');
	}

	/* Treat the value options */
	if (min_run_string) {
		Min_Run_Size = atoi(min_run_string);
		if (Min_Run_Size == 0)
			fatal("bad or zero run size; form is: -r N");
	}
	if (page_width_string) {
		Page_Width = atoi(page_width_string);
		if (Page_Width <= 0)
			fatal("bad or zero page width");
	}
	if (threshold_string) {
		Threshold_Percentage = atoi(threshold_string);
		if ((Threshold_Percentage > 100) || (Threshold_Percentage <= 0))
			fatal("threshold must be between 1 and 100");
	}
	if (output_name) {
		Output_File = fopen(output_name, "w");
		if (Output_File == 0) {
			char *msg = (char *)Malloc(strlen(output_name) + 100);

			sprintf(msg, "cannot open output file `%s'",
				output_name);
			fatal(msg);
			/*NOTREACHED*/
		}
	}

	/* Treat the input-determining options */
	if (is_set_option('i')) {
		/* read input file names from standard input */
		if (argc != 0)
			fatal("-i option conflicts with file arguments");
		get_new_std_input_args(&argc, &argv);
	}
	if (is_set_option('R')) {
		get_new_recursive_args(&argc, &argv);
	}
	/* (argc, argv) now represents new_file* [ / old_file*] */

	/* Here the real work starts */
	Init_Language();

	if (is_set_option('-')) {
		/* Just the lexical scan */
		while (argv[0]) {
			const char *arg = argv[0];
			if (!is_new_old_separator(arg)) {
				Print_Stream(arg);
			}
			argv++;
		}
	}
	else
	if (is_set_option('p')) {
		/* Show percentages */
		read_and_compare_files(argc, argv, 1);
		Show_Percentages();
	} else {
		/* Show runs */
		read_and_compare_files(argc, argv, 1);
		Retrieve_Runs();
		Show_Runs();
	}

	if (is_set_option('M')) {
		/* It is not trivial to plug the leaks, because data structures
		   point to each other, and have to be freed in the proper
		   order. But it is not impossible either. To do, perhaps.
		*/
		ReportMemoryLeaks(stderr);
	}

	return 0;
}
